File: cpp/problem003.cpp
/*
    Copyright (c) 2015 - buyuk.dev@gmail.com
    ----------------------------------------
    Compute the largest prime factor for a 
    given integer n.    
*/

#include <iostream>
#include <vector>
#include <string>
#include <list>
#include <map>
#include <algorithm>
#include <cmath>


#define MAX(a,b) (a<b?b:a)
#define MIN(a,b) (a<b?a:b)



using namespace std;



void sieve(vector<bool>& primes) {
    primes[0] = primes[1] = false;

    for (int step=2; step<primes.size(); step++) {
        for (int i=2*step; i<primes.size(); i+=step) {
            primes[i] = false;
        } 
    } 
}


int main(int argc, char* argv[]) 
{
    ios_base::sync_with_stdio(0);
    
    long number = 0;
    cin >> number;
    
    vector<bool> primes(int(sqrt(number)), true);
    sieve(primes);
    
    for (int i=primes.size()-1; i>=0; --i) {
        if (primes[i] && (number % i == 0)) {
            cout << "result: " << i << endl;
            break;
        }        
    }    
    
    return 0;
}

File: python/problem003.py
#
# Copyright (c) 2016 - buyuk.dev@gmail.com
# ----------------------------------------
# compute largest prime factor for n
#

from math import sqrt


def sieve(lim):

    primes = [True] * (lim + 1)
    primes[0] = primes[1] = False

    for k in range(2, len(primes)):
        for i in range(k * 2, len(primes), k):
            primes[i] = False

    return [n for n, p in enumerate(primes) if p]


if __name__ == '__main__':

    num = 600851475143
    size = int(sqrt(num));

    primes = sieve(size)

    for p in reversed(primes):
        if num % p == 0:
            print "largest prime factor in ", num, " is ", p
            break

